File: src/main/java/com/jozufozu/exnihiloomnia/integration/jei/sieve/SieveRecipeWrapper.java
package com.jozufozu.exnihiloomnia.integration.jei.sieve;

import com.google.common.collect.Lists;
import com.jozufozu.exnihiloomnia.common.registries.WeightedDrop;
import com.jozufozu.exnihiloomnia.common.registries.recipes.SieveRecipe;
import com.jozufozu.exnihiloomnia.integration.jei.JeiPlugin;
import mezz.jei.api.ingredients.IIngredients;
import mezz.jei.api.recipe.IRecipeWrapper;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.Ingredient;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class SieveRecipeWrapper implements IRecipeWrapper
{
    public final Ingredient input;
    public final Map<ItemStack, List<WeightedDrop>> rewards;
    
    public SieveRecipeWrapper(SieveRecipe recipe)
    {
        this.input = recipe.getInput();
        this.rewards = recipe.getOutput().getEquivalencies();
    }
    
    @Override
    public void getIngredients(IIngredients iIngredients)
    {
        List<List<ItemStack>> inputs = new ArrayList<>();
        inputs.add(Lists.newArrayList(input.getMatchingStacks()));
        iIngredients.setInputLists(ItemStack.class, inputs);
        
        iIngredients.setOutputLists(ItemStack.class, JeiPlugin.getRewardsOutput(rewards.keySet(), 45));
    }
}

File: src/main/kotlin/com/jozufozu/exnihiloomnia/integration/jei/sieve/SieveRecipeWrapper.kt
package com.jozufozu.exnihiloomnia.integration.jei.sieve

import com.google.common.collect.Lists
import com.google.common.collect.Maps
import com.jozufozu.exnihiloomnia.common.registries.WeightedDrop
import com.jozufozu.exnihiloomnia.common.registries.recipes.SieveRecipe
import com.jozufozu.exnihiloomnia.integration.jei.JeiPlugin
import mezz.jei.api.ingredients.IIngredients
import mezz.jei.api.ingredients.VanillaTypes
import mezz.jei.api.recipe.IRecipeWrapper
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.Ingredient
import net.minecraftforge.items.ItemHandlerHelper

class SieveRecipeWrapper(recipe: SieveRecipe) : IRecipeWrapper {
    val input: Ingredient = recipe.input
    val rewards: Map<ItemStack, Map<String, List<WeightedDrop>>>

    init {
        val rewards = Maps.newHashMap<ItemStack, HashMap<String, ArrayList<WeightedDrop>>>()

        for (reward in recipe.rewards.outputs) {
            val drop = reward.drop

            if (drop.isEmpty) continue

            val entry = rewards.entries.firstOrNull { drop.isItemEqual(it.key) }

            if (entry == null) {
                rewards[ItemHandlerHelper.copyStackWithSize(drop, 1)] = hashMapOf(Pair(reward.type, arrayListOf(reward)))
            } else {
                entry.value.getOrPut(reward.type, ::ArrayList).add(reward)
            }
        }

        this.rewards = rewards
    }

    override fun getIngredients(iIngredients: IIngredients) {
        val inputs = ArrayList<List<ItemStack>>()
        inputs.add(Lists.newArrayList(*input.matchingStacks))
        iIngredients.setInputLists(VanillaTypes.ITEM, inputs)

        iIngredients.setOutputLists(VanillaTypes.ITEM, JeiPlugin.getRewardsOutput(rewards.keys, 45))
    }
}
